File: chili_native/Services/RuneScapePrices.cs
using System;
using System.Globalization;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace Chili.Services;

/// <summary>A Grand Exchange price for a RuneScape item.</summary>
public sealed record ItemPrice(string Name, long Id, long Price, long Volume, long TimestampMs);

/// <summary>Brief wiki info: an intro extract and an item thumbnail URL.</summary>
public sealed record ItemInfo(string Extract, string? ThumbUrl);

/// <summary>
/// RuneScape price + info engine — C# port of the Flutter <c>runescape_prices.dart</c>.
/// All endpoints are free + keyless:
///   • item search  → RS Wiki MediaWiki opensearch (fuzzy → exact item name)
///   • GE price     → WeirdGloop exchange/history/rs/latest
///   • brief info   → RS Wiki query (intro extract + page thumbnail)
/// </summary>
public sealed class RuneScapePrices
{
    private const string Ua = "CHILI-Home-Copilot/1.0 (rindolf.miaco@gmail.com)";
    private readonly HttpClient _http;

    public RuneScapePrices(HttpClient? http = null)
    {
        _http = http ?? new HttpClient { Timeout = TimeSpan.FromSeconds(12) };
        if (!_http.DefaultRequestHeaders.UserAgent.TryParseAdd(Ua))
            _http.DefaultRequestHeaders.TryAddWithoutValidation("User-Agent", Ua);
    }

    /// <summary>Resolve a fuzzy query to the wiki's exact item name (WeirdGloop is
    /// case/spelling sensitive). Returns null if nothing matched.</summary>
    public async Task<string?> SearchExactName(string query)
    {
        var url = "https://runescape.wiki/api.php?action=opensearch&format=json&limit=5&redirects=resolve&search="
                  + Uri.EscapeDataString(query);
        var json = await _http.GetStringAsync(url).ConfigureAwait(false);
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement; // [term, [names...], [descs...], [urls...]]
        if (root.ValueKind != JsonValueKind.Array || root.GetArrayLength() < 2) return null;
        var names = root[1];
        if (names.ValueKind != JsonValueKind.Array || names.GetArrayLength() == 0) return null;
        return names[0].GetString();
    }

    /// <summary>Latest GE price for an exact item name, or null if unpriced.</summary>
    public async Task<ItemPrice?> PriceByName(string name)
    {
        var url = "https://api.weirdgloop.org/exchange/history/rs/latest?name="
                  + Uri.EscapeDataString(name);
        var json = await _http.GetStringAsync(url).ConfigureAwait(false);
        return ParseWeirdGloopLatest(json, name);
    }

    /// <summary>Search then price in one call (fuzzy query → GE price).</summary>
    public async Task<ItemPrice?> Lookup(string query)
    {
        var exact = await SearchExactName(query).ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(exact)) return null;
        return await PriceByName(exact!).ConfigureAwait(false);
    }

    /// <summary>Download raw bytes (e.g. an item thumbnail) with the shared UA.</summary>
    public Task<byte[]> FetchBytes(string url) => _http.GetByteArrayAsync(url);

    /// <summary>Best-effort intro blurb + thumbnail for an item.</summary>
    public async Task<ItemInfo> WikiInfo(string name)
    {
        var url = "https://runescape.wiki/api.php?action=query&prop=extracts%7Cpageimages"
                  + "&exintro&explaintext&piprop=thumbnail&pithumbsize=96&redirects=1&format=json&titles="
                  + Uri.EscapeDataString(name);
        var json = await _http.GetStringAsync(url).ConfigureAwait(false);
        return ParseWikiInfo(json);
    }

    // ---- pure parsers (unit-testable) -------------------------------------

    /// <summary>Parse a WeirdGloop <c>/latest</c> body. The response is an object
    /// keyed by item name → {id, price, volume, timestamp}.</summary>
    public static ItemPrice? ParseWeirdGloopLatest(string json, string fallbackName)
    {
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        if (root.ValueKind != JsonValueKind.Object) return null;
        foreach (var prop in root.EnumerateObject())
        {
            var v = prop.Value;
            if (v.ValueKind != JsonValueKind.Object) continue;
            // an error payload looks like {"success":false,...}
            if (prop.NameEquals("success") || prop.NameEquals("error")) continue;
            var name = prop.Name.Length > 0 ? prop.Name : fallbackName;
            return new ItemPrice(
                name,
                ReadLong(v, "id"),
                ReadLong(v, "price"),
                ReadLong(v, "volume"),
                ReadTimestampMs(v, "timestamp"));
        }
        return null;
    }

    /// <summary>Parse the wiki query response → extract + thumbnail.</summary>
    public static ItemInfo ParseWikiInfo(string json)
    {
        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("query", out var q) ||
            !q.TryGetProperty("pages", out var pages) ||
            pages.ValueKind != JsonValueKind.Object)
            return new ItemInfo("", null);

        foreach (var page in pages.EnumerateObject())
        {
            var p = page.Value;
            var extract = p.TryGetProperty("extract", out var e) ? e.GetString() ?? "" : "";
            string? thumb = null;
            if (p.TryGetProperty("thumbnail", out var t) &&
                t.TryGetProperty("source", out var s))
                thumb = s.GetString();
            return new ItemInfo(extract, thumb);
        }
        return new ItemInfo("", null);
    }

    private static long ReadLong(JsonElement obj, string key)
    {
        if (!obj.TryGetProperty(key, out var el)) return 0;
        switch (el.ValueKind)
        {
            case JsonValueKind.Number:
                return el.TryGetInt64(out var n) ? n : (long)el.GetDouble();
            case JsonValueKind.String:
                return long.TryParse(el.GetString(), NumberStyles.Any,
                    CultureInfo.InvariantCulture, out var s) ? s : 0;
            default:
                return 0;
        }
    }

    private static long ReadTimestampMs(JsonElement obj, string key)
    {
        if (!obj.TryGetProperty(key, out var el)) return 0;
        if (el.ValueKind == JsonValueKind.Number && el.TryGetInt64(out var ms)) return ms;
        if (el.ValueKind == JsonValueKind.String)
        {
            var str = el.GetString();
            if (long.TryParse(str, out var asMs)) return asMs;
            if (DateTimeOffset.TryParse(str, CultureInfo.InvariantCulture,
                    DateTimeStyles.AssumeUniversal, out var dto))
                return dto.ToUnixTimeMilliseconds();
        }
        return 0;
    }

    // ---- formatting helpers ----------------------------------------------

    /// <summary>Full grouped number, e.g. 85139 → "85,139".</summary>
    public static string FormatGpFull(long n) => n.ToString("N0", CultureInfo.InvariantCulture);

    /// <summary>Abbreviated price, e.g. 85139 → "85.1k", 1234567 → "1.2m".</summary>
    public static string FormatGp(long n)
    {
        double a = Math.Abs(n);
        string sign = n < 0 ? "-" : "";
        if (a >= 1_000_000_000) return $"{sign}{a / 1_000_000_000:0.#}b";
        if (a >= 1_000_000) return $"{sign}{a / 1_000_000:0.#}m";
        if (a >= 1_000) return $"{sign}{a / 1_000:0.#}k";
        return n.ToString(CultureInfo.InvariantCulture);
    }

    /// <summary>Trim a wiki extract to a short blurb at a word boundary.</summary>
    public static string BriefBlurb(string extract, int maxChars = 160)
    {
        if (string.IsNullOrWhiteSpace(extract)) return "";
        var s = extract.Trim();
        if (s.Length <= maxChars) return s;
        var cut = s.LastIndexOf(' ', Math.Min(maxChars, s.Length - 1));
        if (cut < maxChars / 2) cut = maxChars;
        return s.Substring(0, cut).TrimEnd() + "…";
    }
}

File: chili_mobile/lib/src/games/runescape_prices.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// A RuneScape (RS3) Grand Exchange price result (GAME-10).
class ItemPrice {
  const ItemPrice({
    required this.name,
    required this.id,
    required this.price,
    required this.volume,
    required this.timestampMs,
  });

  final String name;
  final String id;
  final int price; // GE guide price in gp
  final int volume; // daily trade volume
  final int timestampMs;
}

/// Parse a MediaWiki opensearch response `[query, [titles], [descs], [urls]]`
/// into the list of item-name suggestions. Pure + tolerant. GAME-10.
List<String> parseOpenSearch(Object? json) {
  if (json is List && json.length >= 2 && json[1] is List) {
    return <String>[
      for (final Object? t in (json[1] as List))
        if (t != null && '$t'.trim().isNotEmpty) '$t'.trim(),
    ];
  }
  return const <String>[];
}

/// Parse a WeirdGloop `/exchange/history/rs/latest` response — a map keyed by
/// item name → {id, price, volume, timestamp}. Returns the first priced entry,
/// or null on an error/empty body. Pure. GAME-10.
ItemPrice? parseWeirdGloopLatest(Map<String, dynamic> json) {
  for (final MapEntry<String, dynamic> e in json.entries) {
    final Object? v = e.value;
    if (v is Map && v['price'] is num) {
      return ItemPrice(
        name: e.key,
        id: '${v['id'] ?? ''}',
        price: (v['price'] as num).toInt(),
        volume: (v['volume'] as num?)?.toInt() ?? 0,
        timestampMs:
            DateTime.tryParse('${v['timestamp']}')?.millisecondsSinceEpoch ?? 0,
      );
    }
  }
  return null;
}

/// Compact gp formatting: 85139 → "85.1K", 1500000 → "1.50M". GAME-10.
String formatGp(int gp) {
  final int a = gp.abs();
  final String sign = gp < 0 ? '-' : '';
  if (a >= 1000000000) return '$sign${(a / 1e9).toStringAsFixed(2)}B';
  if (a >= 1000000) return '$sign${(a / 1e6).toStringAsFixed(2)}M';
  if (a >= 1000) return '$sign${(a / 1e3).toStringAsFixed(1)}K';
  return '$gp';
}

/// Full gp with thousands separators: 85139 → "85,139". GAME-10.
String formatGpFull(int gp) {
  final String digits = gp.abs().toString();
  final StringBuffer b = StringBuffer(gp < 0 ? '-' : '');
  for (int i = 0; i < digits.length; i++) {
    if (i > 0 && (digits.length - i) % 3 == 0) b.write(',');
    b.write(digits[i]);
  }
  return b.toString();
}

/// Looks up RS3 Grand Exchange prices via the RuneScape Wiki opensearch (to
/// resolve a typed query to a real item name) and the WeirdGloop exchange API
/// (for the current GE price). Both are free, keyless, and used by the wiki
/// itself. Injectable [get] for tests. GAME-10.
class RuneScapePrices {
  RuneScapePrices({Future<String> Function(Uri url)? get})
      : _get = get ?? _defaultGet;

  final Future<String> Function(Uri url) _get;

  static const String _ua =
      'CHILI-Home-Copilot/1.0 (desktop game frame price lookup)';

  static Future<String> _defaultGet(Uri url) async {
    final http.Response r =
        await http.get(url, headers: <String, String>{'User-Agent': _ua});
    if (r.statusCode != 200) {
      throw http.ClientException('HTTP ${r.statusCode}', url);
    }
    return r.body;
  }

  /// Item-name suggestions for a typed query (autocomplete).
  Future<List<String>> search(String query) async {
    final String q = query.trim();
    if (q.isEmpty) return const <String>[];
    final Uri url = Uri.parse(
        'https://runescape.wiki/api.php?action=opensearch'
        '&search=${Uri.encodeQueryComponent(q)}'
        '&limit=8&namespace=0&format=json');
    return parseOpenSearch(jsonDecode(await _get(url)));
  }

  /// Current GE price for an exact item name.
  Future<ItemPrice?> priceByName(String name) async {
    final String n = name.trim();
    if (n.isEmpty) return null;
    final Uri url = Uri.parse(
        'https://api.weirdgloop.org/exchange/history/rs/latest'
        '?name=${Uri.encodeQueryComponent(n)}');
    final Object? json = jsonDecode(await _get(url));
    if (json is Map<String, dynamic>) return parseWeirdGloopLatest(json);
    return null;
  }

  /// Resolve a typed query to a priced item: try it as an exact name first,
  /// then fall back to the top opensearch match(es).
  Future<ItemPrice?> lookup(String query) async {
    try {
      final ItemPrice? exact = await priceByName(query);
      if (exact != null) return exact;
    } catch (_) {
      // ignore — fall through to search
    }
    final List<String> names = await search(query);
    for (final String n in names.take(3)) {
      try {
        final ItemPrice? p = await priceByName(n);
        if (p != null) return p;
      } catch (_) {
        // try the next candidate
      }
    }
    return null;
  }
}
